Add abutting directions to excision spheres

nilsvu opened the pull request:
## Proposed changes

This is the information needed to figure out which elements face the excision boundary. Used for interpolations from element faces to a Strahlkorper representing the excision.

### Upgrade instructions

<!--
If this PR makes changes that other people should be aware of when upgrading
their code, describe what they should do between the two UPGRADE INSTRUCTIONS
lines below.
-->
<!-- UPGRADE INSTRUCTIONS -->

<!-- UPGRADE INSTRUCTIONS -->

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `new feature` if appropriate.

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

nilsvu commented:
@gsb76 @markscheel I needed this block neighbor information to get observations on excision boundaries to an (almost!) working state, and I remember we discussed adding this a while back. I hope this helps you as well.

nilsvu commented:
@nikwit ping because I think you are using something like this on a branch as well

wthrowe commented:
Why have you chosen a set here?  It looks like you are using it as a map.  (Iterating over all the pairs looking for a particular first entry.)

nilsvu commented:
I think you're absolutely right that a map makes more sense. Give me a sec to check.

kidder commented on src/Domain/Structure/ExcisionSphere.hpp:
```diff
@@ -27,14 +33,21 @@ class er;
   /// \param center the coordinate center of the excision sphere
   /// in the computational domain.
-  ExcisionSphere(double radius, std::array<double, VolumeDim> center);
+  /// \param block_neighbors the set of blocks that face the excision sphere.
```
face ->  abut

face just means oriented towards, while abut means touching

kidder commented on src/Domain/Structure/ExcisionSphere.hpp:
```diff
@@ -52,13 +65,30 @@ class ExcisionSphere {
+  /// The direction in which the block identified by `block_id` faces the
+  /// excision sphere, or `std::nullopt` if the block does not face the excision
+  /// sphere
+  std::optional<Direction<VolumeDim>> face_direction(
```
face_direction -> abutting_direction

mar-celine commented on src/Domain/Creators/BinaryCompactObject.cpp:
```diff
@@ -511,10 +511,22 @@ Domain<3> BinaryCompactObject::create_domain() const {
       {{"ObjectAExcisionSphere",
         ExcisionSphere<3>{object_A_.inner_radius,
-                          {{object_A_.x_coord, 0.0, 0.0}}}},
+                          {{object_A_.x_coord, 0.0, 0.0}},
```
Maybe a small comment saying "Direction of faces given by [reference to either documentation or code]"

mar-celine reviewed (commented):
Do you know if there's any need to also keep track of the orientation/connectivity of the faces? All of excision boundary directions being lower zeta doesn't determine this and I wonder if knowing which excision boundary faces are neighboring each other is information you might need. Otherwise looks good!

nilsvu committed:
Add abutting directions to excision spheres

This is the information needed to figure out which elements touch
the excision boundary. Used for interpolations from element faces
to a Strahlkorper representing the excision.

nilsvu commented:
I simplified this a lot since @wthrowe is absolutely right that a map is the right representation here. Thanks!

Also squashed in @kidder's and @gsb76's comments.

@gsb76 I believe we don't need orientations and connectivity. My current thought is: a surface integral over the excision boundary decomposes into the sum of the integrals over the individual abutting element faces. So to get the spherical harmonic coefficients we can integrate the spherical harmonics times the data over each abutting element face independently and sum the result. This can be expressed as a projection matrix I believe, but should only need information about the element face. Either way, if I'm wrong we can add the orientation information when we need it.

mar-celine reviewed (dismissed):
Looks good! :+1:

nilsvu commented:
@kidder please check if your comments have been addressed here.
